### PR TITLE
Hook nft liquidity position decaying withdraw fee

### PR DESCRIPTION
**Pull request by johngrantuk**
# Description

This hook/router example uses NFT liquidity positions to track non-fungible metadata, in this case allowing a decaying withdraw fee to be applied to each position.

- Hook is also a custom Router.
- Custom router:
  - is needed so that BPT can be minted to itself rather than msg.sender (Balancer router always transfer to msg.sender)
  - on add, the router keeps the BPT and issues an NFT to the msg.sender
  - NFT ID maps to LP specific data: bpt->to keep track of pool, bptAmount->size of position, startTime->to allow decaying fee to be calculated
  - on remove, the router confirms NFT ownership, uses the BPT to do remove, sends tokens to msg.sender and burns NFT 
- MinimalRouter contract has functions for add/remove proportional with optional receiver param. Idea is this could be reused in other routers.
- Hook functionality:
  - Basic decaying fee that starts at 10% and decays 1% a day
  - Blocks add/remove from any router other than itself
  - Any withdraw fee is paid to pool

Notes/ToDos:
- Must disable unbalanced liquidity in order to take fee. Seems a shame that this affects adds unnecessarily as proportional adds are quite annoying to deal with.
- MinimalRouter inherits from another contract, RouterCommon, that calls VaultGuard(vault) in its constructor. BaseHooks also does this so I got an `Got error: Base constructor arguments given twice.` error. Wasn't sure of best way around this so ended up implementing full hooks interface.
- pkg/pool-hooks/test/foundry/ExitFeeHookExampleStablePool.t.sol (and Weighted variant) - maybe just me but seems like these aren't actually using the hook?

- Initialise is done via normal router so no NFT minted or fee gets applied for this position (and they could never remove) - can update to add initialise to MinimalRouter then implement initialise in NftRouter to issue NFT?
- MinimalRouter only got add/remove kinds of interest at the moment but might be worth creating more full featured as could be used in other hooks?

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Optimization: [ ] gas / [ ] bytecode
- [ ] Documentation or wording changes
- [X] Other - add example hook

## Checklist:

- [ ] The diff is legible and has no extraneous changes
- [ ] Complex code has been commented, including external interfaces
- [ ] Tests have 100% code coverage
- [ ] The base branch is either `main`, or there's a description of how to merge

### CODE REVIEW

File: pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        address,
+        address,
+        TokenConfig[] memory,
+        LiquidityManagement calldata liquidityManagement
+    ) public virtual returns (bool) {
+        // This hook requires donation support to work (see above).
+        if (liquidityManagement.enableDonation == false) {
+            revert PoolDoesNotSupportDonation();
+        }
+
```
Comment (danielmkm) [inline pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol:119]: Should maybe also check that unbalanced operations are disabled.

File: pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    ) public virtual returns (bool, uint256[] memory hookAdjustedAmountsOutRaw) {
+        // We only allow removeLiquidity via the Router/Hook itself so that fee is applied correctly
+        require(router == address(this), "Can't use external router");
+
+        uint256 tokenId = abi.decode(userData, (uint256));
+        hookAdjustedAmountsOutRaw = amountsOutRaw;
+        // Calculate the number of days that have passed since startTime
+        uint256 daysPassed = (block.timestamp - startTime[tokenId]) / 1 days;
+        // Initial fee of 10%
+        uint256 initialFee = 10e16;
```
Comment (danielmkm) [inline pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol:166]: Can you pull these into constants

File: pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        if (daysPassed < 10) {
+            // decreasing fee by 1% per day
+            uint256 currentFee = initialFee - 1e16 * daysPassed;
+            hookAdjustedAmountsOutRaw = _takeFee(pool, amountsOutRaw, currentFee);
```
Comment (danielmkm) [inline pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol:170]: Maybe a view function where I give an `id` and it returns what the remaining fee is.

**Comment by EndymionJkb** [comment]
> MinimalRouter inherits from another contract, RouterCommon, that calls VaultGuard(vault) in its constructor. BaseHooks also does this so I got an Got error: Base constructor arguments given twice. error. Wasn't sure of best way around this so ended up implementing full hooks interface.

Note that this has been fixed as of #925. Other hooks inherit from VaultGuard to get `onlyVault` to protect the hook functions.

File: pkg/pool-hooks/contracts/MinimalRouter.sol [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+} from "@balancer-labs/v3-solidity-utils/contracts/openzeppelin/ReentrancyGuardTransient.sol";
+
+import { RouterCommon } from "@balancer-labs/v3-vault/contracts/RouterCommon.sol";
+
+abstract contract MinimalRouter is RouterCommon, ReentrancyGuardTransient {
+    using Address for address payable;
+
+    /**
+     * Extends AddLiquidityHookParams to include a receiver
+     * @notice Data for the add liquidity hook.
```
Comment (EndymionJkb) [inline pkg/pool-hooks/contracts/MinimalRouter.sol:24]:
```suggestion
     * @notice Data for the add liquidity hook.
     * @dev Extends AddLiquidityHookParams to include a receiver.
```

File: pkg/pool-hooks/contracts/MinimalRouter.sol [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        uint256[] maxAmountsIn;
+        uint256 minBptAmountOut;
+        AddLiquidityKind kind;
+        bool wethIsEth;
+        bytes userData;
+    }
+
+    /**
+     * Extends RemoveLiquidityHookParams to include a receiver
+     * @notice Data for the remove liquidity hook.
```
Comment (EndymionJkb) [inline pkg/pool-hooks/contracts/MinimalRouter.sol:47]:
```suggestion
     * @notice Data for the remove liquidity hook.
     * @dev Extends RemoveLiquidityHookParams to include a receiver.
```

File: pkg/pool-hooks/contracts/MinimalRouter.sol [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                _weth.withdraw(amountOut);
+                ethAmountOut = amountOut;
+            } else {
+                // Transfer the token to the receiver (amountOut).
+                _vault.sendTo(token, params.receiver, amountOut);
+            }
+        }
+
+        // Send ETH to receiver.
+        payable(params.receiver).sendValue(ethAmountOut);
```
Comment (EndymionJkb) [inline pkg/pool-hooks/contracts/MinimalRouter.sol:241]: Could check for 0 as an optimization.

File: pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+pragma solidity ^0.8.24;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { ERC721 } from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import { IPermit2 } from "permit2/src/interfaces/IPermit2.sol";
+
+import { IRouterCommon } from "@balancer-labs/v3-interfaces/contracts/vault/IRouterCommon.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import { IWETH } from "@balancer-labs/v3-interfaces/contracts/solidity-utils/misc/IWETH.sol";
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
```
Comment (EndymionJkb) [inline pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol:12]: Move FixedPoint down under BaseHooks; interfaces should be grouped separate from contract imports.

File: pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+contract NftLiquidityPositionExample is MinimalRouter, ERC721, BaseHooks {
+    using FixedPoint for uint256;
+
+    // This contract uses timestamps to update its withdrawal fee over time.
+    //solhint-disable not-rely-on-time
+
+    // Initial fee of 10%.
+    uint256 public constant INITIAL_FEE_PERCENTAGE = 10e16;
+    uint256 public constant ONE_PERCENT = 1e16;
+    // After this amount of days the fee will be 0%.
```
Comment (EndymionJkb) [inline pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol:36]:
```suggestion
    // After this number of days the fee will be 0%.
```
Sounds more natural to me.

File: pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+    uint256 public constant FULL_DECAY_DAYS = 10;
```
Comment (EndymionJkb) [inline pkg/pool-hooks/contracts/NftLiquidityPositionExample.sol:37]: Maybe `DECAY_PERIOD`? or `DECAY_PERIOD_DAYS`?

**Review by EndymionJkb — APPROVED** [review]
Looks great; just a few minor suggestions.

**Pull request merged**